sympy/printing/pycode.py

```diff
@@ -1062,3 +1062,304 @@ def _print_Function(self, expr):
 
     def _print_Pow(self, expr, rational=False):
         return self._hprint_Pow(expr, rational=rational, sqrt='sympy.sqrt')
+
+
+_not_in_cupy = 'erf erfc factorial gamma loggamma'.split()
+_in_cupy = [(k, v) for k, v in _known_functions_math.items() if k not in _not_in_cupy]
+_known_functions_cupy = dict(_in_cupy, **{
+    'acos': 'arccos',
+    'acosh': 'arccosh',
+    'asin': 'arcsin',
+    'asinh': 'arcsinh',
+    'atan': 'arctan',
+    'atan2': 'arctan2',
+    'atanh': 'arctanh',
+    'exp2': 'exp2',
+    'sign': 'sign',
+    'logaddexp': 'logaddexp',
+    'logaddexp2': 'logaddexp2',
+})
+_known_constants_cupy = {
+    'Exp1': 'e',
+    'Pi': 'pi',
+    'EulerGamma': 'euler_gamma',
+    'NaN': 'nan',
+    'Infinity': 'PINF',
+    'NegativeInfinity': 'NINF'
+}
+
+
+class CuPyPrinter(PythonCodePrinter):
```

Comment:
Is it really necessary to redefine every print method here? Can't this subclass the NumPyPrinter and change the module from `numpy` to `cupy`? It will require a little refactoring in the `NumPyPrinter` to make the `numpy` part a class variable that can be overridden.

Reply:
Yes! I was a bit surprised too when I saw the sheer bulk of print methods. Two not-thought-through-yet possibilities when subclassing NumPyPrinter:

- Call `.replace('numpy', 'cupy')` on the methods, as the api is identical
- Change the NumPyPrinter to only use the `from numpy import *` "syntax", (`np.dot()` -> `dot()`), thus making the strings identical for numpy and cupy.
- Slight alternative to the above is to import both numpy and cupy in their respective printers like `import cupy as xp`, and then use `xp` throughout the code.

Reply:
CC @bjodah. 

I'm a little confused. The printers appear to use `numpy.dot` and so on, but if you look at the source code in lambdify, it just uses `dot`. What is the purpose of using numpy.dot in the printers?

Assuming there is a good reason for it, we should just abstract the `numpy` into a class variable that can be overridden by a CuPyPrinter subclass.